src/scenechange/mod.rs

```diff
@@ -104,237 +136,289 @@ impl SceneChangeDetector {
+      // we assume that we have delta over threshold in back looking deque
+      // and we don't have frames over threshold in forward looking deque
+      if !back_over_tr.is_empty()
+        && forward_over_tr.is_empty()
+        && back_deque.len() > 1
+        && back_over_tr.len() > 1
+      {
+        return true;
       }
+
+      // check for special case
+      // scenecut before flash
+      // longer than max flash length
+      if back_over_tr.is_empty()
+        && forward_over_tr.len() == 1
+        && forward_deque[0].0 > forward_deque[0].1
+      {
+        return true;
+      }
+
+      if !back_over_tr.is_empty() || !forward_over_tr.is_empty() {
+        return false;
+      }
+    }
+
+    scene_score >= scene_threshold
+  }
+
+  /// The fast algorithm detects fast cuts using a raw difference
+  /// in pixel values between the scaled frames.
+  #[hawktracer(fast_scenecut)]
+  fn fast_scenecut(
+    &mut self, frame1: Arc<Frame<T>>, frame2: Arc<Frame<T>>,
+  ) -> ScenecutData {
+    // Downscaling both frames for comparison
+    // Moving scaled frames to buffer
+    if self.frame_buffer.is_empty() {
+      let frame1_scaled = frame1.planes[0].downscale(self.scale_factor);
+      self.frame_buffer.push(frame1_scaled);
+
+      let frame2_scaled = frame2.planes[0].downscale(self.scale_factor);
+      self.frame_buffer.push(frame2_scaled);
+    } else {
+      self.frame_buffer.remove(0);
+      self.frame_buffer.push(frame2.planes[0].downscale(self.scale_factor));
+    }
+
+    let delta =
+      self.delta_in_planes(&self.frame_buffer[0], &self.frame_buffer[1]);
+
+    ScenecutData {
+      intra_cost: self.threshold as f64,
+      threshold: self.threshold as f64,
+      inter_cost: delta as f64,
     }
   }
 
   /// Run a comparison between two frames to determine if they qualify for a scenecut.
   ///
-  /// The standard algorithm uses block intra and inter costs
+  /// Using block intra and inter costs
   /// to determine which method would be more efficient
   /// for coding this frame.
-  ///
-  /// The fast algorithm detects fast cuts using a raw difference
-  /// in pixel values between the frames.
-  /// It does not handle pans well, but the scene flash detection compensates for this
-  /// in many cases.
-  fn has_scenecut<T: Pixel>(
+  #[hawktracer(cost_scenecut)]
+  fn cost_scenecut(
     &self, frame1: Arc<Frame<T>>, frame2: Arc<Frame<T>>, frameno: u64,
     previous_keyframe: u64,
-  ) -> ScenecutResult {
-    if self.fast_mode {
-      let len = frame2.planes[0].cfg.width * frame2.planes[0].cfg.height;
-      let delta = self.delta_in_planes(&frame1.planes[0], &frame2.planes[0]);
-      let threshold = self.threshold * len as u64;
-      ScenecutResult {
-        intra_cost: threshold as f64,
-        threshold: threshold as f64,
-        inter_cost: delta as f64,
-        has_scenecut: delta >= threshold,
-      }
+  ) -> ScenecutData {
+    let frame2_ref2 = Arc::clone(&frame2);
```

Comment:
Is there frame2_ref1 or frame2_ref ?
It maybe slightly confusing if we have 2 at the end if there is no frame2_ref or frame2_ref1. Could simply keep ref itself, 
IIRC it was there earlier, itself, maybe @lu-zero could say if it is good idea to keep frame2_ref itself or not.

Reply:
I don't know why it's done the way it is, I just refactored the code)